### DISCUSSION TITLE
Extracing the embedded preview for RAW files

### DISCUSSION THREAD

**Discussion by mwegithub** [dlemstra/Magick.NET#1006] [Help] [answered]
Hi, this is my first day using Magick.NET so please excuse me if this is already explained and I just did not find it.  I did search GitHub and Google for quite a while.

I'm working mostly with RAW images (NEF, CR2) and I want to extract the **embedded preview JPG** created by the camera: I only need to fallback to processing the full RAW when the embedded preview is too small. This is a matter of _performance_.

I'm currently using LibRaw with C++ and use the `unpack_thumb()` and `dcraw_make_mem_thumb` functions to fetch and access the preview. image. This takes about **0.5 seconds** to complete on my computer for a D850 NEF file.

Environment:
Windows 10 64-bit, 
Magick.NET-Q16-AnyCPU 8.2.0
.Net Core 3.1
C#, Visual Studio 2019

For Magick.NET, I've started by using one of the samples:
```
using (var image = new MagickImage(FileName))
{
    Console.WriteLn("Done");
}
```
This works, but loads the full RAW data and takes about **6 seconds per image**, instead of the 0.5s I'm used to.

I've tried various things, e.g., specifying the width and height in the constructor, providing a `MagickReadSettings`  instance with width/height set, pages, frames....

I could not find anything that would not load the full RAW data but **only** the embedded preview - which is optimized by the camera and optimal for my purposes.

Is there a switch or option which tells IM.NET to load only the embedded preview, but do no other processing (basically a Ping which also reads the JPEG).

**Comment by dlemstra** [comment]
ImageMagick uses libraw and it should be possible to extract the jpg file. Not sure yet what the best soluton be for this. Maybe we could add a define for this to read the thumb and store that as an attribute on the image. Would probably become something like this:

```C#
using (var image = new MagickImage())
{
    image.Settings.SetDefine(new DngReadDefines
    {
        ReadThumbnail = true
    });

    image.Ping(FileName);

    var profile = image.GetProfile("dng:thumbnail");
    File.WriteAllBytes("thumbnail.jpg", profile.GetData());
}
```

**Comment by mwegithub** [comment]
Thank you for looking into this. Much appreciated!

I'm totally new to Magick.NET so I cannot really make suggestions on how to best implement this within the framework.
Reading a RAW image was basically the first test I did with Magick.NET.

If the preview is large enough, it provides an almost instant and usually well-optimized rendition of the RAW data - created using the camera vendors algorithms. For many intents and purposes, the preview is all that's needed. And it's _super fast_.
This is why my current C++ code using LibRaw always tries that first. It falls back to developing the RAW only when there is no preview or the preview is too small (RAW files produced by older cameras).

Maybe this would fit into how Magick.NET deals with _multi-frame / multi-page_ files? That's how Windows WIC does it, when I recall correctly. Treating the 160px thumbnail, the JPG preview and the RAW sensor data as individual frames. Frame 0 = thumbnail, frame 1 = preview, frame 2 = full RAW image... and one can specify which frame to load  via MagickReadSettings...?

There is one thing to consider: **Image Orientation** (rotation).
The embedded preview may have its own EXIF record, with its own EXIF orientation field. Or not. In that case it is assumed that the preview has been rotated to the "neutral" orientation and that it can be used _as-is_. But this is not always the case. Some camera vendors, for some models / firmware version don't rotate the embedded JPEG. In that case the EXIF orientation recorded for the RAW must be used. It's a bit of a mess, really.

I'm new to Magick.NET but I understand that Magick.NET has functions to read EXIF data.
In order to know if and how to rotate the preview image returned, access to EXIF data of the preview and/or the EXIF data of the RAW image is needed. Not sure if reading EXIF is linked to loading the full sensor data or if this can be done separately.

**Comment by dlemstra** [comment, marked as answer]
I decided to use the approach that I described earlier. You can load the thumbnail like this:

```C#
var profile = image.GetProfile("dng:thumbnail");
using (var jpegThumbnail = new MagickImage(profile.GetData()))
{
    // Correct the image orientation
    jpegThumbnail.AutoOrient();
}
```

If the thumbnail contains an exif profile it will be loaded automatically.

**Reply by mwegithub** [reply]
Excellent!
I'm busy with another project right now, but I will check out and check your change next week.
Thank you very much. This enhancement should be helpful for many.
